### PR TITLE
Refactor and improve randperm tests.

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22121 Refactor and improve randperm tests.**
* #22103 Swap detection order in randperm_out_cuda to avoid unnecessary conversion from float when the input is small.
* #22102 Support Half type in randperm.

Differential Revision: [D16153794](https://our.internmc.facebook.com/intern/diff/D16153794)

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -2759,30 +2759,23 @@ def test_nvtx(self):
...
-        self.assertEqual(res1, res2, 0)
-
-        with torch.random.fork_rng(devices=[0]):
-            res1 = torch.randperm(50000, dtype=torch.float, device=cuda)
-        res2 = torch.cuda.FloatTensor()
-        torch.randperm(50000, out=res2, device=cuda)
-        self.assertEqual(res1, res2, 0)
+        # Test core functionality. For small n, randperm is offloaded to CPU instead. For large n, randperm is executed
+        # on GPU.
+        for n in (100, 50000):
```
Comment (SsnL) [inline test/test_cuda.py]: where is the 100000 test case?

Comment (xuhdev) [reply test/test_cuda.py]: Added back.

**Review by soumith — APPROVED** [review]

**Review by yf225 — APPROVED** [review]

**Comment by ailzhang** [comment]
I'm not able to import this PR... cc : @yf225 do you know how to fix?

**Comment by xuhdev** [comment]
@ailzhang I guess the other two dependency PRs must be imported first?

**Comment by gchanan** [comment]
CI failure looks unrelated.

**Pull request closed without merging**